NIFI-2791 Adding 'math' expression language function

JPercivall opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

pvillard31 commented on nifi-commons/nifi-expression-language/src/main/java/org/apache/nifi/attribute/expression/language/evaluation/functions/MathEvaluator.java:
```diff
@@ -0,0 +1,138 @@
+                executionValue = (Number) method.invoke(null);
+            } else if(optionalArg == null) {
+                boolean subjectIsDecimal = subjectValue instanceof Double;
+                Method method = Math.class.getMethod(methodNamedValue, subjectIsDecimal ? double.class : long.class);
```
If I use:
````
${filename:math('exp')}
````
I'll get the following exception:
````
org.apache.nifi.attribute.expression.language.exception.AttributeExpressionLanguageException: Unable to calculate math function value
	at org.apache.nifi.attribute.expression.language.evaluation.functions.MathEvaluator.evaluate(MathEvaluator.java:129) ~[nifi-expression-language-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at org.apache.nifi.attribute.expression.language.Query.evaluate(Query.java:568) ~[nifi-expression-language-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at org.apache.nifi.attribute.expression.language.Query.evaluateExpression(Query.java:386) ~[nifi-expression-language-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at org.apache.nifi.attribute.expression.language.StandardPreparedQuery.evaluateExpressions(StandardPreparedQuery.java:48) ~[nifi-expression-language-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at org.apache.nifi.attribute.expression.language.StandardPropertyValue.evaluateAttributeExpressions(StandardPropertyValue.java:152) ~[nifi-expression-language-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at org.apache.nifi.attribute.expression.language.StandardPropertyValue.evaluateAttributeExpressions(StandardPropertyValue.java:133) ~[nifi-expression-language-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at org.apache.nifi.processors.attributes.UpdateAttribute.executeActions(UpdateAttribute.java:496) ~[na:na]
	at org.apache.nifi.processors.attributes.UpdateAttribute.onTrigger(UpdateAttribute.java:377) ~[na:na]
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) ~[nifi-api-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1064) ~[nifi-framework-core-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:136) [nifi-framework-core-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:47) [nifi-framework-core-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:132) [nifi-framework-core-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_77]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_77]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_77]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_77]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_77]
Caused by: java.lang.NoSuchMethodException: java.lang.Math.exp(long)
	at java.lang.Class.getMethod(Class.java:1786) ~[na:1.8.0_77]
	at org.apache.nifi.attribute.expression.language.evaluation.functions.MathEvaluator.evaluate(MathEvaluator.java:82) ~[nifi-expression-language-1.1.0-SNAPSHOT.jar:1.1.0-SNAPSHOT]
	... 19 common frames omitted
````
I know this is because I didn't use ``toDecimal()`` but I am wondering if you were not expecting to throw the custom exception just few lines after? IMO just throwing the current exception may not be very clear to users trying to understand why it didn't work. I think users may forget to cast arguments and this will probably be the most common mistake, it could be useful to remember this "requirement" in the exception message. Thoughts?

JPercivall replied:
You are totally right. I even hit that exception before and didn't think anything of it. I will re-work it a bit in order to hit the proper error message.

JPercivall commented:
@pvillard31 I addressed your comment and added to the unit test to verify that it properly handles the exceptions. A couple notes:

1: The exception handling is not pretty. That said, I don't think there is a better/more concise way to do it since we're dealing with configurable reflection calls.
2: Normally I would have the exception testing in individual tests with an "@Test(expected = AttributeExpressionLanguageException.class)" but since this is a large test class (that probably should be broken up) that tests all the EL and I wanted to keep it concise.

pvillard31 commented:
+1 thanks @JPercivall ! Merged.
